File: src/main/java/com/stage/teamb/services/EnterpriseService.java
package com.stage.teamb.services;

import com.stage.teamb.dtos.DepartmentDTO;
import com.stage.teamb.dtos.EnterpriseDTO;
import com.stage.teamb.models.Enterprise;

import java.util.List;
import java.util.Optional;

public interface EnterpriseService {


    List<EnterpriseDTO> findAllEnterprises();

    EnterpriseDTO findEnterpriseById(Long id);

    EnterpriseDTO saveEnterprise(EnterpriseDTO enterpriseDTO);

    void deleteEnterpriseById(Long id);
    EnterpriseDTO updateEnterprise(EnterpriseDTO enterpriseDTO);



    List<Enterprise> findAll();

    Optional<Enterprise> findOne(Long id);

    Enterprise saveOne(Enterprise enterprise);

    void deleteOne(Long id);

    Boolean existsById(Long id);

    List<DepartmentDTO> findDepartmentsByEnterpriseId(Long enterpriseId);

    DepartmentDTO associateDepartmentWithEnterprise(Long enterpriseId, DepartmentDTO departmentDTO);

    DepartmentDTO disassociateDepartmentFromEnterprise(Long departmentId);

    EnterpriseDTO findEnterpriseByDepartmentId(Long departmentId);
}

File: src/main/resources/TeamBuilding_frontend-main/src/app/shared/services/enterprise.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AuthService } from './auth.service';

@Injectable({
  providedIn: 'root'
})
export class EnterpriseService {

  private baseUrl = 'http://localhost:8082/api/enterprises';

  constructor(private http: HttpClient, private authService: AuthService) { }

  findAllEnterprises(): Observable<any[]> {
    const request = this.http.get<any[]>(`${this.baseUrl}/find/all`);
    return this.authService.attachTokenToRequest(request);
  }

  findEnterpriseById(id: number): Observable<any> {
    const request = this.http.get<any>(`${this.baseUrl}/get/${id}`);
    return this.authService.attachTokenToRequest(request);
  }

  saveEnterprise(enterpriseDTO: any): Observable<any> {
    const request = this.http.post<any>(`${this.baseUrl}/create`, enterpriseDTO);
    return this.authService.attachTokenToRequest(request);
  }

  updateEnterprise(enterpriseDTO: any): Observable<any> {
    const request = this.http.put<any>(`${this.baseUrl}/update/${enterpriseDTO.id}`, enterpriseDTO);
    return this.authService.attachTokenToRequest(request);
  }

  deleteEnterpriseById(id: number): Observable<void> {
    const request = this.http.delete<void>(`${this.baseUrl}/delete/${id}`);
    return this.authService.attachTokenToRequest(request);
  }

  findDepartmentsByEnterpriseId(enterpriseId: number): Observable<any[]> {
    const request = this.http.get<any[]>(`${this.baseUrl}/departments/${enterpriseId}`);
    return this.authService.attachTokenToRequest(request);
  }

  associateDepartmentWithEnterprise(enterpriseId: number, departmentDTO: any): Observable<any> {
    const request = this.http.post<any>(`${this.baseUrl}/departments/${enterpriseId}`, departmentDTO);
    return this.authService.attachTokenToRequest(request);
  }

  disassociateDepartmentFromEnterprise(departmentId: number): Observable<any> {
    const request = this.http.delete<any>(`${this.baseUrl}/departments/${departmentId}`);
    return this.authService.attachTokenToRequest(request);
  }

  findEnterpriseByDepartmentId(departmentId: number): Observable<any> {
    const request = this.http.get<any>(`${this.baseUrl}/departments/${departmentId}/enterprise`);
    return this.authService.attachTokenToRequest(request);
  }

  
}
